### PR TITLE
[Kernel] Build flash-attn from source

### LINKED ISSUE #8002: [RFC]: Build `vllm-flash-attn` from source
### Motivation.

To use a custom version of PyTorch in vLLM, `vllm-flash-attn` needs to be built with the same version. The easiest way to achieve that is by building it from source during the vLLM build.

### Proposed Change.

We propose 3 different ways of building `vllm-flash-attn` from source: absorbing the package completely, building it as a CMake dependency, or running a nested `pip install`. Currently, alternative 2 is preferred, but we'd like to get feedback on that. I will update this RFC once we decide on an approach.

More details here: https://docs.google.com/document/d/1njmz8NPT3am5gNcjbjzZG1BN-v8wIxWq6vb5QoctuZ0/edit?usp=sharing

### Feedback Period.

_No response_

### CC List.

@WoosukKwon @youkaichao @tlrmchlsmth @bnellnm 

### Any Other Things.

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ProExpertProg**
This PR resolves #8002 and builds vllm-flash-attn from source. This is required for using torch nightly.

This PR relies on the new CMake-based build system in vllm-flash-attn. 

To make installation smoother, this process uses an additional CMake install step for putting the .so targets in the correct place instead of using `CMAKE_LIBRARY_OUTPUT_DIRECTORY` during build.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -457,6 +468,8 @@ def _read_requirements(filename: str) -> List[str]:
 
 if _is_cuda() or _is_hip():
     ext_modules.append(CMakeExtension(name="vllm._moe_C"))
+    ext_modules.append(
+        CMakeExtension(name="vllm.vllm_flash_attn.vllm_flash_attn_c"))
```
Comment (bnellnm) [inline setup.py:472]: This might only be applicable in CUDA mode.

Comment (bnellnm) [inline setup.py:472]: This might only be applicable in CUDA mode.  I don't know if it has ever been tested with HIP/AMD

File: CMakeLists.txt [hunk 2]
```diff
@@ -323,3 +324,48 @@ if(VLLM_GPU_LANG STREQUAL "CUDA" OR VLLM_GPU_LANG STREQUAL "HIP")
...
+  set(VLLM_FLASH_ATTN_SRC_DIR $ENV{VLLM_FLASH_ATTN_SRC_DIR})
+endif()
+
+if(VLLM_FLASH_ATTN_SRC_DIR)
+  FetchContent_Declare(flash-attention SOURCE_DIR ${VLLM_FLASH_ATTN_SRC_DIR})
+else()
+  # TODO(luka) use vllm-project version
+  FetchContent_Declare(
+          flash-attention
+          GIT_REPOSITORY https://github.com/neuralmagic/flash-attention.git
```
Comment (youkaichao) [inline CMakeLists.txt:349]: fix this? and do we use the `main` branch or pin the commit?

Comment (ProExpertProg) [reply CMakeLists.txt:349]: I think we pin the commit. This can only be fixed once the vllm-flash-attn PR #19 is merged

**Review by youkaichao — APPROVED** [review]
LGTM in general! can you fix the tests on format?

File: CMakeLists.txt (hunk 2 above)
Comment (youkaichao) [reply CMakeLists.txt:349]: the link is already merged?

Comment (ProExpertProg) [reply CMakeLists.txt:349]: Sorry I meant [this PR](https://github.com/vllm-project/flash-attention/pull/19) on the vllm-flash-attn repo

File: CMakeLists.txt [hunk 3]
```diff
@@ -323,13 +325,64 @@ define_gpu_extension_target(
...
+# The vllm-flash-attn install rules are nested under vllm to make sure the library gets installed in the correct place.
+# To only install vllm-flash-attn, use --component vllm-flash-attn_c.
+# If no component is specified, vllm-flash-attn is still installed.
 
-if(VLLM_GPU_LANG STREQUAL "CUDA" OR VLLM_GPU_LANG STREQUAL "HIP")
-  message(STATUS "Enabling C extension.")
-  add_dependencies(default _C)
-
-  message(STATUS "Enabling moe extension.")
-  add_dependencies(default _moe_C)
```
Comment (ProExpertProg) [inline CMakeLists.txt:333]: It turns out we already have a return statement above that exits unless we're on hip or CUDA, so this if statement was redundant

**Review by bnellnm — APPROVED** [review]

File: .github/workflows/scripts/build.sh [hunk 4]
```diff
@@ -15,5 +15,6 @@ $python_executable -m pip install -r requirements-cuda.txt
 export MAX_JOBS=1
 # Make sure release wheels are built for the following architectures
 export TORCH_CUDA_ARCH_LIST="7.0 7.5 8.0 8.6 8.9 9.0+PTX"
+export VLLM_FA_CMAKE_GPU_ARCHES="80-real;90-real"
```
Comment (mgoin) [inline .github/workflows/scripts/build.sh:18]: What is the "real" for? Does this cover PTX for 8.6/8.9?

**Review by mgoin — APPROVED** [review]
LGTM so just marked ready

File: .github/workflows/scripts/build.sh (hunk 4 above)
Comment (ProExpertProg) [reply .github/workflows/scripts/build.sh:18]: No I think we specifically decided no ptx, that's how it was build before. The reason this is in a different format is it doesn't go through the torch parsing, but gets used directly by CMake. If ptx was needed, we'd say 80-virtual here.

**Comment by youkaichao** [comment]
>  [2024-09-20T18:01:49Z] (VllmWorkerProcess pid=184) INFO 09-20 11:01:49 selector.py:259] Cannot use FlashAttention-2 backend because the vllm_flash_attn package is not found. `pip install vllm-flash-attn` for better performance.

@ProExpertProg you need to update the selection logic.

**Comment by ProExpertProg** [comment]
@youkaichao I updated the import statement and error message. But we could also just get rid of the optional import, as vllm-flash-attn should always get included/built - let me know what you think

**Comment by mgoin** [comment]
We won't build it for non-CUDA backends, so I think it's good practice to keep it as an optional import

**Comment by youkaichao** [comment]
thanks for the great job! @ProExpertProg

File: CMakeLists.txt [hunk 5]
```diff
@@ -70,19 +84,6 @@ endif()
...
-# be kept in sync.
-#
-# The `default` target makes direct use of cmake easier since knowledge
-# of which extensions are supported has been factored in, e.g.
-#
-# mkdir build && cd build
-# cmake -G Ninja -DVLLM_PYTHON_EXECUTABLE=`which python3` -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=../vllm ..
-# cmake --build . --target default
-#
-add_custom_target(default)
```
Comment (youkaichao) [inline CMakeLists.txt:85]: @ProExpertProg  looks like this fails cpu ci build. is it necessary to delete these lines?

Comment (ProExpertProg) [reply CMakeLists.txt:85]: No, we should just remove the dependency. Explained in the comment in #8722, addressed in #8723.

**Pull request merged**